### PR TITLE
Feature/cadsify

### PR DESCRIPTION
**Pull request by EddyCMWF**
Still working on this, but creating the PR for some nice visuals and to keep track of conversations

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](http://claassistant.ecmwf.int:80/ecmwf-projects/cads-toolbox?pullRequest=14) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](http://claassistant.ecmwf.int:80/ecmwf-projects/cads-toolbox?pullRequest=14) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: JamesVarndell<br/>:x: EddyCMWF<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://claassistant.ecmwf.int:80/check/ecmwf-projects/cads-toolbox?pullRequest=14) it.</sub>

File: tests/test_10_inputs_transform.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+def test_transform(request_args):
+
+    ct.config.USE_CACHE = False
+    remote = ct.catalogue.retrieve(*request_args)
+    transformed = inputs_transform.transform(remote, xr.Dataset)
+    assert type(transformed) == xr.Dataset
+
+    local = remote.data
+    transformed = inputs_transform.transform(local, xr.Dataset)
+    assert type(transformed) == xr.Dataset
```
Comment (JamesVarndell) [inline tests/test_10_inputs_transform.py:59]:
```suggestion
    assert isinstance(transformed, xr.Dataset)
```

File: tests/test_10_inputs_transform.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import pandas as pd
+import pytest
+import xarray as xr
+
+import cads_toolbox as ct
+from cads_toolbox import inputs_transform
+
+
+def dummy_ndarray_function(value: np.ndarray):
+    assert type(value) == np.ndarray
```
Comment (JamesVarndell) [inline tests/test_10_inputs_transform.py:11]:
```suggestion
    assert isinstance(value, np.ndarray)
```

File: tests/test_10_inputs_transform.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def dummy_dataset_function(value: xr.Dataset):
+    assert type(value) == xr.Dataset
```
Comment (JamesVarndell) [inline tests/test_10_inputs_transform.py:15]:
```suggestion
    assert isinstance(value, xr.Dataset)
```

File: tests/test_10_inputs_transform.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def dummy_dataarray_function(value: xr.DataArray):
+    assert type(value) == xr.DataArray
```
Comment (JamesVarndell) [inline tests/test_10_inputs_transform.py:19]:
```suggestion
    assert isinstance(value, xr.DataArray)
```

File: tests/test_10_inputs_transform.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def dummy_dataframe_function(value: pd.DataFrame):
+    assert type(value) == pd.DataFrame
```
Comment (JamesVarndell) [inline tests/test_10_inputs_transform.py:23]:
```suggestion
    assert isinstance(value, pd.DataFrame)
```

**Review by JamesVarndell — CHANGES_REQUESTED** [review]
These changes look great! I've made a few specific suggestions, but I also have a couple of more general suggestions:

1. Consider renaming `inputs_transform` to use a private namespace (e.g. `_inputs_transform`, as no user should be accessing this module directly. If you do make this change, you can also remove leading underscores from the functions within `_inputs_transform`.
2. You've got nice tests for transforming kwargs in functions with arg types set, but none for the magic named input variables (e.g. "dataarray"). It would be nice to add a test for this case.

File: cads_toolbox/cadsify.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    "dataarray": xr.DataArray,
+    "dataset": xr.Dataset,
+    "data": np.ndarray,
+}
+
+
+def transform(thing, kwarg_type):
+    if type(thing) == Remote:
+        thing = thing.data
+    return emohawk.transform(thing, kwarg_type)
```
[the remark this answers is not in the document]
Comment (EddyCMWF) [reply cads_toolbox/cadsify.py]: Update made

File: cads_toolbox/cadsify.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        for arg, name in zip(args, signature.parameters):
+            kwargs[name] = arg
+        # transform kwargs if necessary
+        for key, value in [(k,v) for k,v in kwargs.items() if k in mapping]:
+            kwarg_types = ensure_iterable(mapping[key])
+            for kwarg_type in kwarg_types:
+                if kwarg_type is type(value):
+                    break
+            else:
+                kwargs[key] = transform(value, kwarg_types[0])
```
[the remark this answers is not in the document]
Comment (EddyCMWF) [reply cads_toolbox/cadsify.py]: done

File: cads_toolbox/cadsify.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        result = function(**kwargs)
+        try:
+            return emohawk.open(result)
+        except:
+            # TODO: Make not bare
+            return result
+
+    @wraps(function)
+    def wrapper(*args, **kwargs):
```
[the remark this answers is not in the document]
Comment (EddyCMWF) [reply cads_toolbox/cadsify.py]: I'd prefer not to as then it is clear that `wrapper` is just the wrapper decorator and _wrapper handles the kwarg_types

**Review by JamesVarndell — APPROVED** [review]

**Pull request merged**